File: src/main/java/com/matchFit/post/repository/PostRepository.java
package com.matchFit.post.repository;

import com.matchFit.post.entity.Post;
import com.matchFit.post.entity.Status;
import java.time.LocalDateTime;
import java.util.List;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.transaction.annotation.Transactional;

public interface PostRepository extends JpaRepository<Post, Long>, PostRepositoryCustom {

    List<Post> findAllByDateBetween(LocalDateTime start, LocalDateTime end);

    List<Post> findByUserIdOrderByCreatedAtDesc(Long userId);

    @Query("select p.id from Post p where p.date < :now and p.status = :open")
    List<Long> findExpiredCandidateIds(
            @Param("now") LocalDateTime now,
            @Param("open") Status open
    );

    @Modifying
    @Transactional
    @Query("update Post p set p.status = :expired where p.date < :now and p.status = :open")
    int markExpired(
            @Param("now") LocalDateTime now,
            @Param("expired") Status expired,
            @Param("open") Status open
    );
}

File: src/main/kotlin/com/matchFit/post/repository/PostRepository.kt
package com.matchFit.post.repository

import com.matchFit.post.entity.Post
import com.matchFit.post.entity.Status
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import java.time.LocalDateTime

interface PostRepository : JpaRepository<Post, Long>, PostRepositoryCustom {

    fun findAllByDateBetween(start: LocalDateTime, end: LocalDateTime): List<Post>

    fun findByUserIdOrderByCreatedAtDesc(userId: Long): List<Post>

    @Modifying
    @Query("update Post p set p.status = :expired where p.date < :now and p.status = :open")
    fun markExpired(
        @Param("now") now: LocalDateTime,
        @Param("expired") expired: Status,
        @Param("open") open: Status
    ): Int
}
